### PR TITLE
Port the rest of our geometry and utility functions from the old code 

### LINKED ISSUE #25: Port the rest of our geometry and utility functions from the old code
We will need all of our geometry and utility functions from the old code, including the different shapes, lines, and all the functions used to find intercepts etc. These should be brought into the new code.

### PR DESCRIPTION
**Pull request by jordanzliu**
resolves #25

### CODE REVIEW

File: src/thunderbots/software/geom/rect.h [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+#include "geom/point.h"
+
+/**
+ * A rectangle.
+ */
+class Rect final
+{
+   public:
+    /**
+     * Creates a new Rect from two corners.
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/rect.h:14]: Should have javadoc style comments (for the whole file)

File: src/thunderbots/software/geom/shapes.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#include "geom/point.h"
+#include "geom/rect.h"
+
+class Line final
+{
+   public:
+    Point first;
+    Point second;
+
+    /**
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/shapes.h:13]: Should be javadoc style comments (for the whole file)

File: src/thunderbots/software/geom/util.cpp [hunk 3]
```diff
@@ -0,0 +1,965 @@
...
+    const Vector &src, const Vector &p1, const Vector &p2,
+    const std::vector<Vector> &obstacles, const double &radius)
+{
+    std::vector<std::pair<Vector, Angle>> ret;
+
+    const Angle offangle = (p1 - src).orientation();
+    if (collinear(src, p1, p2))
+    {
+        // std::cerr << "geom: collinear " << src << " " << p1 << " " << p2 <<
+        // std::endl;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: We should have better error handling here. We should decide if when the lines are colinear, if we just take the circles that intersect the line, or if we return an "empty" result. We can use std::optional for the empty result.

My vote would be for returning the circles that intersect the line.

File: src/thunderbots/software/geom/util.cpp [hunk 4]
```diff
@@ -0,0 +1,965 @@
...
+        if (diff.len() < radius)
+        {
+            return std::make_pair(bestshot, Angle::zero());
+        }
+        const Angle cent   = (diff.orientation() - offangle).angleMod();
+        const Angle span   = Angle::asin(radius / diff.len());
+        const Angle range1 = cent - span;
+        const Angle range2 = cent + span;
+
+        /*
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: Dead code. Unless it's useful and should be uncommented, delete it.

File: src/thunderbots/software/geom/util.cpp [hunk 5]
```diff
@@ -0,0 +1,965 @@
...
+        {
+            ans.push_back(lineIntersection(a, b, segA, segB));
+        }
+    }
+    return ans;
+}
+
+Vector vectorRectIntersect(const Rect &r, const Vector &vecA, const Vector &vecB)
+{
+    /*std::cout << vecA << vecB << r.neCorner() << r.swCorner();
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: Dead code. Unless it's useful and should be uncommented, delete it.

File: src/thunderbots/software/geom/util.cpp [hunk 6]
```diff
@@ -0,0 +1,965 @@
...
+        ret.set(ret.x(), miny);
+    }
+    else if (p.y() > maxy)
+    {
+        ret.set(ret.x(), maxy);
+    }
+    return ret;
+}
+
+// TODO: does this work? unit test
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: Scary `TODO`. Might as well add some unit tests

File: src/thunderbots/software/geom/util.cpp [hunk 7]
```diff
@@ -0,0 +1,965 @@
...
+    // TODO figure out why this is asserting
+    // assert(std::abs((d - c).cross(b - a)) > EPS);
+    if (std::fabs((d - c).cross(b - a)) < EPS)
+    {
+        // LOG_WARN(u8"Cross product problem again");
+    }
+
+    return a + (a - c).cross(d - c) / (d - c).cross(b - a) * (b - a);
+
+    //	// testing new code from
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: Dead code. Unless it's useful and should be uncommented, delete it.

File: src/thunderbots/software/geom/util.cpp [hunk 8]
```diff
@@ -0,0 +1,965 @@
...
+    Vector au = a / a.len();
+    Vector c  = au + b / b.len();
+    // use similar triangle
+    return c * (radius / std::fabs(au.cross(c)));
+}
+
+Vector calcBlockCone(
+    const Vector &a, const Vector &b, const Vector &p, const double &radius)
+{
+    /*
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: Dead code. Unless it's useful and should be uncommented, delete it.

File: src/thunderbots/software/geom/util.h [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+
+/**
+ * Gets the side that is connected to the vertex with index provided
+ * and also connected to the vertex with the next index.
+ */
+template <size_t N>
+Seg getSide(const Poly<N> &poly, unsigned int i);
+
+/**
+ * Checks if 3 points are collinear.
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.h:107]: Should be javadoc style comments (for the whole file)

File: src/thunderbots/software/test/geom/util.cpp [hunk 10]
```diff
@@ -0,0 +1,794 @@
...
+TEST(GeomUtilTest, test_contains_triangle_point)
+{
+    // this triangle lies in the first quatren of the field, we can rota
+    Point p1(0, 0);
+    Point p2((std::rand() % 100) / 100, 0);
+    Point p3((std::rand() % 100) / 100, (std::rand() % 100) / 100);
+    Point p((std::rand() % 100) / 100, (std::rand() % 100) / 100);
+    bool expected_val = true;
+
+    /* i don't know what is going on here, this part seems to not work very well
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/geom/util.cpp]: Dead code. Unless it's useful and should be uncommented, delete it.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Done feedback round 1. Search-and-replace is good for converting doxygen comments to javadoc FYI.

**Comment by MathewMacDougall** [comment]
Also don't forget to edit the description of the PR to show what Issue this resolves
eg. `resolves #<issue number>`

File: src/thunderbots/software/geom/rect.h (hunk 1 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/rect.h:14]: DOne

File: src/thunderbots/software/geom/util.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: Could you clarify what you mean by that? I think it would make sense to return a vector containing the direction of the line and a zero radian or EPS radian angle if the line does not intersect any obstacles.

File: src/thunderbots/software/geom/util.cpp (hunk 6 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: This is actually tested, forgot to remove todo

File: src/thunderbots/software/geom/rect.h [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+inline constexpr Point Rect::seCorner() const
+{
+    return min_corner + Point(diag.x(), 0);
+}
+
+inline constexpr Point Rect::operator[](unsigned int pos) const
+{
+    return (pos % 4) == 0
```
Comment (henryymliu) [inline src/thunderbots/software/geom/rect.h]: This should be refactored with a switch statement on {0, 1, 2, 3} (I think?). I have no idea what usage of this would use any index > 3.
Exception should be thrown if out-of-index.

File: src/thunderbots/software/geom/shapes.h [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    inline explicit constexpr Circle() : radius(0)
+    {
+    }
+
+    inline explicit constexpr Circle(const Point& origin, double r)
+        : origin(origin), radius(r)
+    {
+    }
+
+    // TODO: these shapes could use more operators and functions (eg area), and hashes
```
Comment (henryymliu) [inline src/thunderbots/software/geom/shapes.h]: Might be a bit of scope creep, but can this TODO be addressed? Shouldn't be too difficult to add an area and hash function (function of radius + centre).

File: src/thunderbots/software/geom/shapes.h [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    inline constexpr bool operator==(const Circle& p)
+    {
+        return this->origin == p.origin && this->radius == p.radius;
+    }
+
+    inline constexpr bool operator!=(const Circle& p)
+    {
+        return !(*this == p);
```
Comment (henryymliu) [inline src/thunderbots/software/geom/shapes.h]: wonder why the != operator has much different behaviour than ==...

File: src/thunderbots/software/geom/util.cpp [hunk 14]
```diff
@@ -0,0 +1,914 @@
...
+{
+    return distsq(ray.start, ray.dir) < EPS2;
+}
+
+double len(const Seg &seg)
+{
+    return dist(seg.start, seg.end);
+}
+
+double len(const Line &line)
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: Should maybe remove this.

File: src/thunderbots/software/geom/util.cpp [hunk 15]
```diff
@@ -0,0 +1,914 @@
...
+    (void)line;  // unused
+    return std::numeric_limits<double>::infinity();
+}
+
+double lensq(const Seg &seg)
+{
+    return distsq(seg.start, seg.end);
+}
+
+double lensq(const Line &line)
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp:118]: ditto

File: src/thunderbots/software/geom/util.cpp [hunk 16]
```diff
@@ -0,0 +1,914 @@
...
+               0 &&
+           sign((second.end - second.start).cross(first.start - second.start)) *
+                   sign((second.end - second.start).cross(first.end - second.start)) <=
+               0;
+}
+
+template <size_t N>
+Vector getVertex(const Poly<N> &poly, unsigned int i)
+{
+    return poly[i % N];
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: Instead of modulo, should probably clamp index to [0, N].

File: src/thunderbots/software/geom/util.cpp [hunk 17]
```diff
@@ -0,0 +1,914 @@
...
+    const Vector &a, const Vector &b, const Vector &c, const Vector &d)
+{
+    return std::abs((d - c).cross(b - a)) > EPS;
+}
+
+std::vector<Point> lineIntersection(const Seg &a, const Seg &b)
+{
+    if (std::fabs((b.end - b.start).cross(a.end - a.start)) < EPS)
+    {
+        // LOG_WARN(u8"Cross product problem again in new function");
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: remove?

File: src/thunderbots/software/geom/util.cpp [hunk 18]
```diff
@@ -0,0 +1,914 @@
...
+                                  (b.end - b.start).cross(a.end - a.start) *
+                                  (a.end - a.start)};
+}
+
+// ported code
+Vector lineIntersection(
+    const Vector &a, const Vector &b, const Vector &c, const Vector &d)
+{
+    // TODO figure out why this is asserting
+    // assert(std::abs((d - c).cross(b - a)) > EPS);
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: welp

Comment (henryymliu) [reply src/thunderbots/software/geom/util.cpp]: this whole function is sketch

File: src/thunderbots/software/geom/util.cpp [hunk 19]
```diff
@@ -0,0 +1,914 @@
...
+// ported code
+Vector calcBlockOtherRay(const Vector &a, const Vector &c, const Vector &g)
+{
+    return reflect(c - a, g - c);  // this, and the next two instances, were
+                                   // changed from a - c since reflect() was
+                                   // fixed
+}
+
+// ported code
+bool goalieBlockGoalPost(
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: No idea what it does, but it probably belong in this file...

File: src/thunderbots/software/geom/util.cpp [hunk 20]
```diff
@@ -0,0 +1,914 @@
...
+    const Vector &a, const Vector &b, const Vector &c, const Vector &g)
+{
+    Vector R = reflect(c - a, g - c);
+    return fabs(R.cross(b - c)) < EPS;
+}
+
+// ported code
+Vector calcBlockConeDefender(
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: ditto

File: src/thunderbots/software/geom/util.cpp [hunk 21]
```diff
@@ -0,0 +1,914 @@
...
+    // get normal to line
+    n = x1 - x0;
+    n = n.norm();
+
+    v = p - x0;
+
+    return n.dot(v);
+}
+
+Vector segmentNearLine(Vector a0, Vector a1, Vector b0, Vector b1)
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp:804]: no idea what this does

File: src/thunderbots/software/geom/util.h [hunk 22]
```diff
@@ -0,0 +1,486 @@
...
+ * param[in] dir is the direction of the vector
+ *
+ * param[in] p is the point is question
+ */
+bool pointInFrontVector(Point offset, Point dir, Point p);
+
+/**
+ * Returns true if v2 is clockwise relative to v1
+ */
+// TODO: this should work but hasn't been properly tested
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.h]: -.-

**Review by henryymliu — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/geom/rect.h (hunk 11 above)
Comment (henryymliu) [reply src/thunderbots/software/geom/rect.h]: *out of range

Comment (jordanzliu) [reply src/thunderbots/software/geom/rect.h]: yes

File: src/thunderbots/software/geom/shapes.h (hunk 12 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/shapes.h]: done

File: src/thunderbots/software/geom/shapes.h (hunk 13 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/shapes.h]: fixed

File: src/thunderbots/software/geom/util.cpp (hunk 14 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: yes

File: src/thunderbots/software/geom/util.cpp (hunk 18 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: fixed

File: src/thunderbots/software/geom/util.cpp (hunk 21 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp:804]: nearest point on line segment to line

File: src/thunderbots/software/geom/util.h [hunk 23]
```diff
@@ -0,0 +1,486 @@
...
+ */
+Point calcBlockConeDefender(
+    const Point &a, const Point &b, const Point &c, const Point &g, const double &r);
+
+/*
+ * Ported code from CM geom util
+ */
+
+/**
+ * returns perpendicular offset from line x0-x1 to point p
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.h:401]: This comment and all those below could be converted to javadoc style

File: src/thunderbots/software/geom/util.cpp [hunk 24]
```diff
@@ -665,18 +660,35 @@ std::vector<Point> lineIntersection(const Seg &a, const Seg &b)
...
+    double x3 = line2.start.x();
+    double y3 = line2.start.y();
+    double x4 = line2.end.x();
+    double y4 = line2.end.y();
+
+    double denom = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4);
+    if (denom == 0)
     {
-        // LOG_WARN(u8"Cross product problem again");
+        throw std::out_of_range("fuck");
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: Although I would find running into this exception pretty funny, let's use something more descriptive and less profane.

Preferably, this is where it would be good to return an `std:optional<Point>`, since we are not "guaranteed" an intersection if the lines are parallel.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/geom/util.cpp (hunk 21 above)
Comment (henryymliu) [reply src/thunderbots/software/geom/util.cpp:804]: Ah, missed the comment in the header  :P

**Review by henryymliu — CHANGES_REQUESTED** [review]
FYI there's still a few existing comments that haven't been addressed yet

File: src/thunderbots/software/geom/util.cpp (hunk 24 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: Changed to use std::optional

File: src/thunderbots/software/geom/util.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: According to the existing documentation, the function actually returns pairs of direction vectors and angles from said vectors that that are clear of obstacles between src and the line segment between p1 and p2 (ie. directions and angles that aren't blocked between a ball and the goal). I think it would make sense to return the direction and a zero angle if the line is not intersected by an obstacle.

File: src/thunderbots/software/geom/util.h (hunk 22 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.h]: This is actually tested, I think the todo is removed now

File: src/thunderbots/software/geom/util.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/geom/util.cpp]: bump

File: src/thunderbots/software/test/geom/util.cpp (hunk 10 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/test/geom/util.cpp]: bump

File: src/thunderbots/software/geom/util.cpp [hunk 25]
```diff
@@ -0,0 +1,957 @@
...
+                                   // changed from a - c since reflect() was
+                                   // fixed
+}
+
+// ported code
+Vector calcBlockConeDefender(const Vector &a, const Vector &b, const Vector &c,
+                             const Vector &g, const double &r)
+{
+    Vector R = reflect(c - a, g - c);
+    // std::cout << (R + c) << std::endl;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: not using this print statement so lets remove it.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/geom/util.cpp (hunk 15 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/geom/util.cpp:118]: bump

File: src/thunderbots/software/geom/util.cpp (hunk 16 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/geom/util.cpp]: bump

File: src/thunderbots/software/geom/util.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/geom/util.cpp]: Ah yea I was misunderstanding the purpose of the function. According to the comment if the lines are colinear an empty vector is returned. I think what you said makes sense (return the line and Angle::zero()), so if you make that change make sure to update the comment.

**Comment by MathewMacDougall** [comment]
Bumped some comments I might not have tagged properly last time. Mostly just removing dead code. Otherwise, I think there's just the comments in geom/util.cpp and it's good to go.

File: src/thunderbots/software/test/geom/util.cpp (hunk 10 above)
Comment (jordanzliu) [reply src/thunderbots/software/test/geom/util.cpp]: done

File: src/thunderbots/software/geom/util.cpp (hunk 5 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: done

File: src/thunderbots/software/geom/util.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: updated documentation

**Review by MathewMacDougall — DISMISSED** [review]
LGTM

**Comment by MathewMacDougall** [comment]
@henrymliu final comments?

File: src/thunderbots/software/geom/util.cpp [hunk 26]
```diff
@@ -0,0 +1,939 @@
...
+
+Vector reflect(const Vector &a, const Vector &b, const Vector &p)
+{
+    // Make a as origin.
+    // Rotate by 90 degrees, does not matter which direction?
+    Vector n = (b - a).rotate(Angle::quarter());
+    return a + reflect(p - a, n);
+}
+
+// ported code
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: Are all the following `// ported code` comments needed? Can be removed otherwise.

File: src/thunderbots/software/geom/util.cpp (hunk 16 above)
Comment (henryymliu) [reply src/thunderbots/software/geom/util.cpp]: bump

File: src/thunderbots/software/geom/shapes.h (hunk 12 above)
Comment (henryymliu) [reply src/thunderbots/software/geom/shapes.h]: Didn't notice, but this TODO comment can be removed now.

File: src/thunderbots/software/geom/util.cpp [hunk 27]
```diff
@@ -0,0 +1,939 @@
...
+template <size_t N>
+Vector getVertex(const Poly<N> &poly, unsigned int i)
+{
+    return poly[i % N];
+}
+
+template <size_t N>
+void setVertex(Poly<N> &poly, unsigned int i, const Vector &v)
+{
+    poly[i % N] = v;
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: Same as above.

File: src/thunderbots/software/geom/util.cpp [hunk 28]
```diff
@@ -0,0 +1,939 @@
...
+    if (lenA < lenB)
+    {
+        return segA;
+    }
+    return segB;
+}
+
+namespace
+{
+// this function is never used
```
Comment (henryymliu) [inline src/thunderbots/software/geom/util.cpp]: Comment can be removed.

**Review by henryymliu — CHANGES_REQUESTED** [review]
Really sorry to delay this further, but just a few minor things left. Once that's done then it LGTM.

**Comment by MathewMacDougall** [comment]
I also just merged my STP branch, so you'll want to update from `master`. Shouldn't be any conflicts.

File: src/thunderbots/software/geom/util.cpp (hunk 26 above)
Comment (jordanzliu) [reply src/thunderbots/software/geom/util.cpp]: removed

**Review by henryymliu — APPROVED** [review]
LGTM 👍
